### PR TITLE
KAFKA-17850: fix leaking internal exception in state manager

### PR DESCRIPTION
**Pull request by sebastienviale**
Following the [KIP-1033](https://cwiki.apache.org/confluence/display/KAFKA/KIP-1033%3A+Add+Kafka+Streams+exception+handler+for+exceptions+occurring+during+processing) a FailedProcessingException is passed to the Streams-specific uncaught exception handler.

The goal of the PR is to unwrap a FailedProcessingException into a StreamsException when an exception occurs during the flushing or closing of a store

### CODE REVIEW

**Comment by cadonna** [comment]
Could you please add a description to the PR so that it is clear what it changes. Linking to the JIRA ticket is actually not really needed since the JIRA ticket number can be found in the title.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java [hunk 1]
```diff
@@ -538,13 +539,16 @@ public void flush() {
...
+                        if (exception instanceof FailedProcessingException)
+                            firstException = new StreamsException(exception.getCause());
+                        else if (exception instanceof StreamsException)
                             firstException = exception;
                         else
                             firstException = new ProcessorStateException(
                                 format("%sFailed to flush state store %s", logPrefix, store.name()), exception);
                     }
-                    log.error("Failed to flush state store {}: ", store.name(), exception);
+                    log.error("Failed to flush state store {}: ", store.name(), exception.getCause());
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java]: Why did you change this? If a `StreamsException` that is NOT a `FailedProcessingException` is thrown, it would be perfectly fine to log the exception.
You could do something like:

```java
log.error("Failed to flush state store {}: ", store.name(), exception instanceof FailedProcessingException ? exception.getCause() : exception);
``` 

or maybe it is cleaner to copy the log message inside the `if (firstException == null) {`.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java [hunk 2]
```diff
@@ -538,13 +539,16 @@ public void flush() {
                 } catch (final RuntimeException exception) {
                     if (firstException == null) {
                         // do NOT wrap the error if it is actually caused by Streams itself
-                        if (exception instanceof StreamsException)
+                        // In case of FailedProcessingException Do not keep the failed processing exception in the stack trace
+                        if (exception instanceof FailedProcessingException)
+                            firstException = new StreamsException(exception.getCause());
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java]: Before we introduced the `FailedProcessingException` we did not wrap the exception into a plain `StreamsException` but we wrapped it into a `ProcessorStateException` in the `else`-branch. We should keep that behavior.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java [hunk 3]
```diff
@@ -538,13 +539,16 @@ public void flush() {
                 } catch (final RuntimeException exception) {
                     if (firstException == null) {
                         // do NOT wrap the error if it is actually caused by Streams itself
-                        if (exception instanceof StreamsException)
+                        // In case of FailedProcessingException Do not keep the failed processing exception in the stack trace
+                        if (exception instanceof FailedProcessingException)
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:543]: Could you please add unit tests to `ProcessorStateManagerTest` that verify the correct behavior?
You should also verify the behavior when a processing error handler that continues instead of fails is set.

**Review by cadonna — COMMENTED** [review]
Thanks for the PR, @sebastienviale !

Here my comments.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java (hunk 1 above)
Comment (sebastienviale) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java]: I copied the log.error in the `if (firstException == null)`:
`log.error("Failed to flush cache of store {}: ", store.name(), firstException);`
else 
`log.error("Failed to flush cache of store {}: ", store.name(), exception);`

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java (hunk 2 above)
Comment (sebastienviale) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java]: In case of FailedProcessingException, wrap it into a ProcessorStateException

```
 if (exception instanceof FailedProcessingException) {
    firstException = new ProcessorStateException(exception.getCause());
}
```

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java (hunk 3 above)
Comment (sebastienviale) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:543]: I added tests for flush and close in ProcessorStateManagerTest

For flushCache I added tests in ProcessingExceptionHandlerIntegrationTest

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java [hunk 4]
```diff
@@ -538,13 +539,18 @@ public void flush() {
...
+                            firstException = new ProcessorStateException(exception.getCause());
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java]: IMO, we should keep it exactly the same as before the `FailedProcessingException` was introduced:
```suggestion
                            firstException = new ProcessorStateException(
                                format("%sFailed to flush state store %s", logPrefix, store.name()), exception.getCause());
```

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorStateManagerTest.java [hunk 5]
```diff
@@ -771,6 +772,38 @@ public void close() {
...
+        final MockKeyValueStore stateStore = new MockKeyValueStore(persistentStoreName, true) {
+            @Override
+            public void flush() {
+                throw new FailedProcessingException("processor", exception);
+            }
+        };
+        stateManager.registerStore(stateStore, stateStore.stateRestoreCallback, null);
+
+        final ProcessorStateException thrown = assertThrows(ProcessorStateException.class, stateManager::flush);
+        assertEquals(exception, thrown.getCause());
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorStateManagerTest.java:790]: Please also verify the message and that the stack trace does not contain the `FailedProcessingException`.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorStateManagerTest.java [hunk 6]
```diff
@@ -771,6 +772,38 @@ public void close() {
...
+        final MockKeyValueStore stateStore = new MockKeyValueStore(persistentStoreName, true) {
+            @Override
+            public void close() {
+                throw new FailedProcessingException("processor", exception);
+            }
+        };
+        stateManager.registerStore(stateStore, stateStore.stateRestoreCallback, null);
+
+        final ProcessorStateException thrown = assertThrows(ProcessorStateException.class, stateManager::close);
+        assertEquals(exception, thrown.getCause());
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorStateManagerTest.java:829]: Same as above

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java (hunk 3 above)
Comment (cadonna) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:543]: Why could you not also add a test for `flushCache()` to the unit tests?

**Review by cadonna — COMMENTED** [review]
Thanks for the updates, @sebastienviale !

Here my feedback.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java (hunk 4 above)
Comment (sebastienviale) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java]: done

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorStateManagerTest.java (hunk 5 above)
Comment (sebastienviale) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorStateManagerTest.java:790]: done

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorStateManagerTest.java (hunk 6 above)
Comment (sebastienviale) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorStateManagerTest.java:829]: done

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java (hunk 3 above)
Comment (sebastienviale) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:543]: done
good catch, I was not able to call flushCache() from the actual MockKeyValueStore.
So i added a new MockCacheKeyValueStore that also implements the CachedStateStore

**Comment by sebastienviale** [comment]
I fixed all your comments @cadonna

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorStateManagerTest.java (hunk 5 above)
Comment (cadonna) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorStateManagerTest.java:790]: Could you also verify the stacktrace here? Something like:
```java
assertFalse(Arrays.stream(thrown.getStackTrace()).anyMatch(element -> element.getClassName().contains(FailedProcessingException.class.getSimpleName())));
```
IMO, we should not pollute the stack trace with internal exceptions we only need to skip the exception handler.

This also applies to the other added tests.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java [hunk 7]
```diff
@@ -573,16 +581,23 @@ public void flushCache() {
                 } catch (final RuntimeException exception) {
                     if (firstException == null) {
                         // do NOT wrap the error if it is actually caused by Streams itself
-                        if (exception instanceof StreamsException) {
+                        // In case of FailedProcessingException Do not keep the failed processing exception in the stack trace
+                        if (exception instanceof FailedProcessingException) {
+                            firstException = new ProcessorStateException(
+                                format("%sFailed to flush state store %s", logPrefix, store.name()),
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java]: This should be `%sFailed to flush cache of store %s` as on line 593.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java [hunk 8]
```diff
@@ -618,13 +633,20 @@ public void close() throws ProcessorStateException {
                 } catch (final RuntimeException exception) {
                     if (firstException == null) {
                         // do NOT wrap the error if it is actually caused by Streams itself
-                        if (exception instanceof StreamsException)
+                        // In case of FailedProcessingException Do not keep the failed processing exception in the stack trace
+                        if (exception instanceof FailedProcessingException)
+                            firstException = new ProcessorStateException(
+                                format("%sFailed to flush state store %s", logPrefix, store.name()),
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java]: This should be `%sFailed to close state store %s` as on line 645

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java [hunk 9]
```diff
@@ -538,13 +539,20 @@ public void flush() {
...
+                                exception.getCause());
+                        else if (exception instanceof StreamsException)
                             firstException = exception;
                         else
                             firstException = new ProcessorStateException(
                                 format("%sFailed to flush state store %s", logPrefix, store.name()), exception);
+                        log.error("Failed to flush cache of store {}: ", store.name(), firstException);
+                    } else {
+                        log.error("Failed to flush cache of store {}: ", store.name(), exception);
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java]: These two log messages should be `Failed to flush state store {}: ` as the removed `log.error()` says.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java [hunk 10]
```diff
@@ -618,13 +633,20 @@ public void close() throws ProcessorStateException {
...
+                                format("%sFailed to flush state store %s", logPrefix, store.name()),
+                                exception.getCause());
+                        else if (exception instanceof StreamsException)
                             firstException = exception;
                         else
                             firstException = new ProcessorStateException(
                                 format("%sFailed to close state store %s", logPrefix, store.name()), exception);
+                        log.error("Failed to flush cache of store {}: ", store.name(), firstException);
+                    } else {
+                        log.error("Failed to flush cache of store {}: ", store.name(), exception);
```
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java]: These two log messages should be `Failed to close state store {}:` as the removed log.error() says.

**Review by cadonna — COMMENTED** [review]
@sebastienviale , great that you found a way to also test `flushCache` in a unit test.

I still have some comments. But after you handled those, I think we are ready to merge.

File: streams/src/test/java/org/apache/kafka/test/MockCacheKeyValueStore.java [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.test;
+
+import org.apache.kafka.streams.state.internals.CacheFlushListener;
+import org.apache.kafka.streams.state.internals.CachedStateStore;
+
+public class MockCacheKeyValueStore extends MockKeyValueStore implements CachedStateStore<Object, Object> {
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/test/MockCacheKeyValueStore.java]: Could you please call this class `MockCachedKeyValueStore`?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorStateManagerTest.java (hunk 6 above)
Comment (sebastienviale) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorStateManagerTest.java:829]: done

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorStateManagerTest.java (hunk 5 above)
Comment (sebastienviale) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorStateManagerTest.java:790]: done  for all  added tests

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java (hunk 7 above)
Comment (sebastienviale) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java]: done, sorry it was a bad copy / paste

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java (hunk 8 above)
Comment (sebastienviale) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java]: done, sorry it was a bad copy / paste

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java (hunk 9 above)
Comment (sebastienviale) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java]: done, sorry it was a bad copy / paste

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java (hunk 10 above)
Comment (sebastienviale) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java]: done, sorry it was a bad copy / paste

File: streams/src/test/java/org/apache/kafka/test/MockCacheKeyValueStore.java (hunk 11 above)
Comment (sebastienviale) [reply streams/src/test/java/org/apache/kafka/test/MockCacheKeyValueStore.java]: rename done

**Comment by sebastienviale** [comment]
@cadonna I hope this is ok now !
thanks for your feedbacks

**Review by cadonna — APPROVED** [review]
Thanks, @sebastienviale !

LGTM!

**Comment by sebastienviale** [comment]
@cadonna one test was failing after changing the return type to ProcessorStateException. I fixed it.

**Comment by cadonna** [comment]
Thank you for fixing the test!
Restarted the builds.

**Comment by mjsax** [comment]
Should we cherry-pick this to `3.9` branch?

**Comment by cadonna** [comment]
Yeah, sorry forgot about that.

**Comment by cadonna** [comment]
Cherry-picked to `3.9`
